### ISSUE TITLE
Create a t/X0 map for Geant and FLUKA

### ISSUE THREAD

**Issue by asalzburger** [asalzburger/flag#8]
Use the same cylinder source and create 

- [ ]  t/X0 vs eta material plot for Geant
- [ ] t/X0 vs eta material plot for FLUKA (as described in #7)
- [ ] bring them into a form to overlay them and compare them statistically

You can upload your results here.

**Comment by H4m5t3r** [comment]
![fluka_acts_comparison](https://github.com/asalzburger/flag/assets/56361643/2814d53c-5519-45b2-a7a5-a6e76aa31143)

ACTS in orange and FLUKA in blue.

**Comment by H4m5t3r** [comment]
![fluka_geant4_comparison_conversion](https://github.com/asalzburger/flag/assets/56361643/1bd5d9cc-7ab9-4ef3-a8b6-0571179a3f8b)

A plot where the FLUKA plot is the converted Geant4 geometry.

Geant4 in orange and FLUKA in blue.

I'm checking if I can find anything that could be causing the strange results.

**Comment by asalzburger** [comment]
Cool stuff though. I am confident we find the problem.

**Comment by H4m5t3r** [comment]
Me and Mika had a look at the FLUKA input files generated by pyg4ometry last Friday. The strange results were caused by an error in the definition of the silicon tube and its surroundings. I will check if there is something wrong with the way I defined the geometry in pyg4ometry that could be confusing the converter.

**Comment by H4m5t3r** [comment]
It seems like I just made a mistake somewhere in my code. After I looked through the code and automated the whole conversion and comparison process the converted Geant4 output values are now exactly the same as in the output from Mika's premade example. So the conversion works.

![Geant4_FLUKA_comparison](https://github.com/asalzburger/flag/assets/56361643/1dfb3690-eefe-4321-af93-e4267c9524bf)

**Comment by asalzburger** [comment]
That is excellent, can you expand this to a multi layered cylinder of different materials at next?

**Comment by H4m5t3r** [comment]
Yes. Any wishes for the materials? And should the silicon be the the outermost or innermost material or is it irrelevant?

**Comment by H4m5t3r** [comment]
Multiple cylinders and multiple layers implemented. Everything seems normal so far.

![Screenshot from 2023-07-24 13-39-29](https://github.com/asalzburger/flag/assets/56361643/29d77883-f2c5-4f90-9612-79b2a5dacfa8)
![Screenshot from 2023-07-24 13-39-46](https://github.com/asalzburger/flag/assets/56361643/f2222604-b5a7-4b99-88bd-dff9addaeabb)

**Comment by asalzburger** [comment]
Fantastic - does the loop to fluka still work?

**Comment by H4m5t3r** [comment]
Yes, the loop still works. One thing I noticed is that when the number of layers is increased to over 6 the graphs start to diverge. Should they still be the same? This only seems to happen when the number of layers is increased. The number of cylinders does not affect this.

1 layer
![1](https://github.com/asalzburger/flag/assets/56361643/f5c3cb43-ab94-493c-985f-e0c67015c012)

6 layers
![6](https://github.com/asalzburge

8 layers
![8](https://github.com/asalzburger/flag/assets/56361643/6d3acfe2-bac5-47b0-b5bd-b50ba714831d)

16 layers
![16](https://github.com/asalzburger/flag/assets/56361643/9f07483e-f072-477f-a1bf-948417c1a996)

**Comment by asalzburger** [comment]
Interesting, can it be that the cylinders intersect if they are too many?

**Comment by H4m5t3r** [comment]
Maybe. FLUKA's graphical interface Flair gives warnings that the cylinders are touching each other. Mika mentioned something about that shapes should not be set to start exactly where another shape ends in FLUKA so I can try adding 0.1 between the cylinders and post an update here.

**Comment by asalzburger** [comment]
Good idea, please keep a thin separation between the cylinders.

**Comment by H4m5t3r** [comment]
The warnings about that the cylinders are touching are gone after I added a small gap between them, but the divergence still remains. There are still some warnings about that the surrounding space is touching the edges of the cylinders (in FLUKA) so I can try to see if I can make some changes so that all the warnings disappear.

![Screenshot from 2023-07-27 14-25-50](https://github.com/asalzburger/flag/assets/56361643/7e5289b7-801d-4897-91a3-9fc5e86e967a)

**Comment by asalzburger** [comment]
Could it be that one of the materials is not recognized?

**Comment by H4m5t3r** [comment]
Fixing the warnings about the polygons touching each other didn't help, so there doesn't seem to be any problem in the conversion process.  I tried fixing the warnings manually to see if it would have any effect, but the results were the same.

**Comment by H4m5t3r** [comment]
It could be something with the materials in the ACTS material recording. When I launch it the beginning it warns that the materials are referenced before they are declared.

G4GDML: Reading 'silicon_cylinder.gdml'...
G4GDML: VALIDATION ERROR! ID attribute 'G4_U' is referenced but was never declared at line: 73
G4GDML: VALIDATION ERROR! ID attribute 'G4_Ag' is referenced but was never declared at line: 73
G4GDML: VALIDATION ERROR! ID attribute 'G4_Au' is referenced but was never declared at line: 73
G4GDML: VALIDATION ERROR! ID attribute 'G4_Si' is referenced but was never declared at line: 73
G4GDML: VALIDATION ERROR! ID attribute 'G4_Galactic' is referenced but was never declared at line: 73
G4GDML: Reading definitions...
G4GDML: Reading materials...
G4GDML: Reading solids...
G4GDML: Reading structure...
G4GDML: Reading setup...
G4GDML: Reading 'silicon_cylinder.gdml' done!

Is there aything I can do about this?

**Comment by asalzburger** [comment]
You could in fact do the cross check by hand: the radiation lengths of materials can be obtained from tables, e.g the NIST database. Together with the thickness of the cylinders you can calculate it by hand - the shape comes from the incident angle.

**Comment by asalzburger** [comment]
Exactly, basically, with different layers for different material sand segments in z, phi, we can adjust the granularity of the description accordingly.

**Comment by asalzburger** [comment]
Validation:
- material is additive, i.e. if you cross one layer of Silicon of 1 mm at normal incident, you should see `1 [mm] / X0 [mm]`
- consequently if you go through `n` layers of Silicon of each 1 mm thickness at normal incident, the amount of material you see is multiplied by `n`

Cross-check:
- [ ] Check if Geant4 is ok: make one layer of silicon, make 1 layers, make 6 layers, make n layers, and see if the amount of material scales by `n`
- [ ] If that's good, change to different materials 
- [ ] Cross-check the same with FLUKA
- [ ] If that's good, change to different materials in FLUKA

**Comment by asalzburger** [comment]
Next step afterwards:
- [ ] Introduce segmentation in `phi` and `z`

**Comment by H4m5t3r** [comment]
The Geant4 material recordings seem to scale correctly. At higher number of layers the scaling factor seems to become a bit smaller, but it still looks alright.

Geant4 1 layer
![G4_1](https://github.com/asalzburger/flag/assets/56361643/d0fab7d2-3ffa-486c-a612-51831b648f7c)

Geant4 2 layers
![G4_2](https://github.com/asalzburger/flag/assets/56361643/2ec17d09-f84b-4414-8adc-bcf604ac80df)

Geant4 6 layers
![G4_6](https://github.com/asalzburger/flag/assets/56361643/76796391-5e10-430d-9e1f-5f21b04dd74c)

Geant4 16 layers
![G4_16](https://github.com/asalzburger/flag/assets/56361643/c30da9ca-875b-41f9-81f2-36303d7364a1)

Geant4 32 layers
![G4_32](https://github.com/asalzburger/flag/assets/56361643/0a0029c6-92d8-4832-ad0a-5dfcd2ae8a08)

**Comment by H4m5t3r** [comment]
In the FLUKA simulations something seems to go wrong with 6+ layers. The results scale correctly for up to 5 layers, but after that the results are just the same as for 5 layers.

FLUKA 1 layer
![FLUKA_1](https://github.com/asalzburger/flag/assets/56361643/46642415-4615-4256-9871-d4e9a8d2e863)

FLUKA 2 layers
![FLUKA_2](https://github.com/asalzburger/flag/assets/56361643/70567b58-a86a-434d-b9d7-bb697247f891)

FLUKA 2 layers
![FLUKA_3](https://github.com/asalzburger/flag/assets/56361643/70624bae-fb67-4ff4-8791-61e7e3c546ed)

FLUKA 4 layers
![FLUKA_4](https://github.com/asalzburger/flag/assets/56361643/e4fbc1d3-324b-4c7b-a71b-be0999e1f166)

FLUKA 5 layers
![FLUKA_5](https://github.com/asalzburger/flag/assets/56361643/7baca165-cba8-4f11-ba85-0d08a64b7fa2)

FLUKA 6 layers
![FLUKA_6](https://github.com/asalzburger/flag/assets/56361643/81172767-4765-4cfc-a09a-6744c4d94abe)

FLUKA 16 layers
![FLUKA_16](https://github.com/asalzburger/flag/assets/56361643/21cbf20d-2805-451a-9537-829e268c0361)

FLUKA 32 layers
![FLUKA_32](https://github.com/asalzburger/flag/assets/56361643/11e1f531-8ee3-4d49-8055-200b4d346c76)

**Comment by H4m5t3r** [comment]
Here are also the results from when 4 different materials (silicon, lithium, aluminum and titanium) were used for the Geant4 material recordings.

Geant4 4 layers 4 materials
![G4_4_4](https://github.com/asalzburger/flag/assets/56361643/c36f7049-109a-42c0-ab73-25657941d867)

Geant4 8 layers 4 materials
![G4_4_8](https://github.com/asalzburger/flag/assets/56361643/4eed7e40-c1bc-4611-9063-98dc04eb642b)

Geant4 8 layers 4 materials
![G4_4_12](https://github.com/asalzburger/flag/assets/56361643/600cfabd-68d8-41f1-8bb0-5d8fc8dd28a4)

Geant4 8 layers 4 materials
![G4_4_16](https://github.com/asalzburger/flag/assets/56361643/d816cdd8-3fa8-4324-bf87-43f5779292a6)

Geant4 8 layers 4 materials
![G4_4_20](https://github.com/asalzburger/flag/assets/56361643/f74ee6c3-1ce4-45e6-b5c2-1a9dc7c68cb8)

**Comment by asalzburger** [comment]
Is the code/produced files somewhere that I can look at it? I would like to see if there's anything odd with the Fluka conversion.

**Comment by asalzburger** [comment]
Also, just to make sure we have checked Mika's concern: if you make n cylinders, can you - as a cross-check - make the additional cylinders at decreasing radius and not increasing ones?

**Comment by asalzburger** [comment]
Hey @H4m5t3r - looking through your code I see that the FLUKA conversion always uses the GDML file at input - did you try to convert to FLUKA on the fly when you create the Geant4 geometry? Probably the file conversion is the problem for the 6 layer cutoff?

**Comment by asalzburger** [comment]
Did you check this?
